File: skellig-test-step-processing-rmq/src/main/java/org/skellig/teststep/processor/rmq/model/RmqTestStep.java
package org.skellig.teststep.processor.rmq.model;

import org.skellig.teststep.processing.model.TestStep;
import org.skellig.teststep.processing.model.TestStepExecutionType;
import org.skellig.teststep.processing.model.ValidationDetails;

import java.util.Map;
import java.util.Optional;

public class RmqTestStep extends TestStep {

    private String sendTo;
    private String receiveFrom;
    private String respondTo;
    private String routingKey;

    protected RmqTestStep(String id, String name, TestStepExecutionType execution, int timeout, int delay,
                          Map<String, Object> variables, Object testData, ValidationDetails validationDetails,
                          String sendTo, String receiveFrom, String respondTo, String routingKey) {
        super(id, name, execution, timeout, delay, variables, testData, validationDetails);
        this.sendTo = sendTo;
        this.receiveFrom = receiveFrom;
        this.respondTo = respondTo;
        this.routingKey = routingKey;
    }

    public Optional<String> getSendTo() {
        return Optional.ofNullable(sendTo);
    }

    public Optional<String> getReceiveFrom() {
        return Optional.ofNullable(receiveFrom);
    }

    public Optional<String> getRespondTo() {
        return Optional.ofNullable(respondTo);
    }

    public String getRoutingKey() {
        return routingKey;
    }

    public static class Builder extends TestStep.Builder {

        private String sendTo;
        private String receiveFrom;
        private String respondTo;
        private String routingKey;

        public Builder withSendTo(String sendTo) {
            this.sendTo = sendTo;
            return this;
        }

        public Builder withReceiveFrom(String receiveFrom) {
            this.receiveFrom = receiveFrom;
            return this;
        }

        public Builder withRespondTo(String respondTo) {
            this.respondTo = respondTo;
            return this;
        }

        public Builder withRoutingKey(String routingKey) {
            this.routingKey = routingKey;
            return this;
        }

        @Override
        public RmqTestStep build() {
            return new RmqTestStep(id, name, execution, timeout, delay, variables, testData, validationDetails,
                    sendTo, receiveFrom, respondTo, routingKey);
        }
    }
}

File: skellig-test-step-processing-rmq/src/main/kotlin/org/skellig/teststep/processor/rmq/model/RmqTestStep.kt
package org.skellig.teststep.processor.rmq.model

import org.skellig.teststep.processing.model.ScenarioStateUpdater
import org.skellig.teststep.processing.model.TestStepExecutionType
import org.skellig.teststep.processing.model.ValidationNode

/**
 * Represents a test step that can have details about where to read data from RMQ queues, where to send it and respond upon receive.
 * Usually, individual properties [sendTo] and [readFrom] are used, or a combination of:
 * 1) [sendTo] and [readFrom] - Defines 2 execution steps where to send a message ([testData]) first and read a response after the message is sent.
 * 2) [readFrom] and [respondTo] - Defines 2 execution steps where to read a message from and respond with a defined [testData] to another queue.
 * The [testData] property is mandatory [sendTo] or [respondTo] queues are set.
 *
 * @property sendTo The set of queues to send data to.
 * @property readFrom The set of queues to read data from.
 * @property respondTo The set of queues to respond to.
 * @property routingKey - RMQ routing key for responses.
 * @property properties - RMQ queue properties.
 */
open class RmqTestStep protected constructor(id: String?,
                                             name: String?,
                                             execution: TestStepExecutionType?,
                                             timeout: Int,
                                             delay: Int,
                                             attempts: Int,
                                             values: Map<String, Any?>?,
                                             testData: Any?,
                                             validationDetails: ValidationNode?,
                                             scenarioStateUpdaters: List<ScenarioStateUpdater>?,
                                             val sendTo: Set<String>?,
                                             val readFrom: Set<String>?,
                                             val respondTo: Set<String>?,
                                             routingKey: String?,
                                             properties: Map<String, Any?>?)
    : BaseRmqTestStep(id, name!!, execution, timeout, delay, attempts,
                      values, testData, validationDetails, scenarioStateUpdaters, routingKey, properties) {

    override fun toString(): String {
        return super.toString() +
                (sendTo?.let { "sendTo = $sendTo\n" } ?: "") +
                (readFrom?.let { "readFrom = $readFrom\n" } ?: "") +
                (respondTo?.let { "respondTo = $respondTo\n" } ?: "")
    }

    class Builder : BaseRmqTestStep.Builder<RmqTestStep>() {

        private var sendTo: Set<String>? = null
        private var readFrom: Set<String>? = null
        private var respondTo: Set<String>? = null

        fun sendTo(sendTo: Set<String>?) = apply {
            this.sendTo = sendTo
        }

        fun readFrom(readFrom: Set<String>?) = apply {
            this.readFrom = readFrom
        }

        fun respondTo(respondTo: Set<String>?) = apply {
            this.respondTo = respondTo
        }

        override fun build(): RmqTestStep {
            return RmqTestStep(id, name, execution, timeout, delay, attempts, values, testData, validationDetails,
                scenarioStateUpdaters, sendTo, readFrom, respondTo, routingKey, properties)
        }
    }
}